examples/main_test.go

```diff
@@ -0,0 +1,33 @@
+package examples
+
+import (
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+)
+
+func TestMain(m *testing.M) {
+	// Ensure we use a locally built provider
+	awsProviderPath, err := exec.LookPath("pulumi-resource-aws")
+	if err != nil {
+		log.Println("AWS provider not found in path!")
+		os.Exit(1)
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		log.Println("Unable to get working directory!")
+		os.Exit(1)
+	}
+	desiredPath := filepath.Join(cwd, "..", "bin", "pulumi-resource-aws")
+
+	if awsProviderPath != desiredPath {
+		log.Println("AWS provider not built locally!")
```

Comment:
I think we would need to tell the user exactly what command to run to remedy this problem. Would it be `make provider`? 

Our makefiles are not in a good place. I look at:

```
provider: tfgen install_plugins
	(cd provider && go build $(PULUMI_PROVIDER_BUILD_PARALLELISM) -o $(WORKING_DIR)/bin/$(PROVIDER) -ldflags "-X $(PROJECT)/$(VERSION_PATH)=$(VERSION) -X github.com/hashicorp/terraform-provider-aws/version.ProviderVersion=$(VERSION)" $(PROJECT)/$(PROVIDER_PATH)/cmd/$(PROVIDER))
```

The problem with this is that tfgen would be re-running if we called `make provider` all the time because it doesn't know how to be incremental. I wonder if we could lift this clause into its own make target, and then automatically run it from the tests:

```
(cd provider && go build $(PULUMI_PROVIDER_BUILD_PARALLELISM) -o $(WORKING_DIR)/bin/$(PROVIDER) -ldflags "-X $(PROJECT)/$(VERSION_PATH)=$(VERSION) -X github.com/hashicorp/terraform-provider-aws/version.ProviderVersion=$(VERSION)" $(PROJECT)/$(PROVIDER_PATH)/cmd/$(PROVIDER))
```

The `go build` command is adequately incremental and fast enough on modern hardware. If we auto-run it from tests, this would mean that beyond guaranteeing that the provider was built locally, it also makes sure that if you made any changes to the provider, it gets rebuilt so the changes are included. You won't get into the situations where the test gives an unexpected result because you forgot to rebuild the provider.

Reply:
I've added the step for building the provider to the `test_main`. I used `make provider -o tfgen -o install_plugins`, which would run whatever `go` command `make` would run, avoiding `tfgen` and `install_plugins`. Let me know if that doesn't seem right, happy to change it to run the actual `go` command itself.

I also added an env var to disable this behaviour so that CI runs are not affected, since they have a separate job for running the build. https://github.com/pulumi/ci-mgmt/pull/717 sets this env var in the CI files. We should merge this first if we want to retain the CI behaviour.

Reply:
After rereading your comment I realised that you propose adding a new make target, which is more sensible. I added that in https://github.com/pulumi/ci-mgmt/pull/718